File: scripts/getMatchesData.js
// Connect to DB
mongoose.connect(config.db.host + ':' + config.db.port + '/' + config.db.name);
var request = require('sync-request');
var Match = require('../models/match');
var _ = require('lodash');
var async = require('async');
var STARTTIME = 1427865900;
var CURRENTTIME = STARTTIME;
var TIME_JUMP = 5 * 60; //5 minutes in seconds
var SLEEP_TIME = 2 * 60 * 1000; //2 minute in milliseconds
var RUN = true;

var get_game_ids_queue = [];
var get_game_data_queue = [];

var STATUS_CODES = {
	OK: 200,
	BAD_REQUEST: 400,
	UNAUTHORIZED: 401,
	RATE_EXCEEDED: 429,
	INTERNAL_SERVER_ERROR: 500,
	SERVICE_UNAVAILABLE: 503
};

var FUTURE_TIME_TEXT = 'beginDate is invalid';

async.series([
	function (cb) {
		Match.remove({}, function (err) {
			console.log('Match collection cleared');
			cb();
		});
	},
	function (cb) {
		function magicks = function () {
			if (get_game_ids_queue.length < 100) { //Only enqueue when we have space
				config.riot.regions.forEach(function (region) {
					var call = {query: region.endpoint + 'api/lol/' + region.name + '/v4.1/game/ids?beginDate=' + CURRENTTIME + '&api_key=' + config.riot.apiKey, 
								region: region.name,
								region_endpoint: region.endpoint};
					if (get_game_ids_queue.indexOf(call) !== -1) return;
					get_game_ids_queue.push(call);
					get_game_ids_queue = _.uniq(get_game_ids_queue, 'query');
				});
				CURRENTTIME += TIME_JUMP;
			}
			
			var value = get_game_ids_queue.pop();
			var res = request('GET', value.query);

			if (res.statusCode === STATUS_CODES.OK) {
				var body = JSON.parse(res.body);
				body.forEach(function (gameId) {
					var call = value.region_endpoint + 'api/lol/' + value.region + '/v2.2/match/' + gameId + '?includeTimeline=false&api_token=' + config.riot.apiKey;
					get_game_data_queue.push({query: call,
												region: value.region});
				});
			} else if (res.statusCode === STATUS_CODES.RATE_EXCEEDED {
				get_game_ids_queue.unshift(value); //Return value to top of queue
				setTimeout(magicks, SLEEP_TIME); //Sleep for 2 seconds and restart the process
				return;
			}
		}
		setTimeout(magicks, 100);
	}
	], function (err) {
		if (err) console.log(err);
		else console.log('Process finished successfully')
		process.exit(0);
	})

File: scripts/getMatchesData.py
from pymongo import MongoClient
import requests
import time
from collections import deque
import json

STARTTIME = 1427865900
CURRENTTIME = STARTTIME
TIME_JUMP = 5 * 60 #5 minutes in seconds
SLEEP_TIME = 2 * 60  #2 minutes in seconds
APITOKEN = '040f1207-45a5-45f7-bdc5-e41356ce433e'

client = MongoClient('localhost', 27017)
db = client['lol-bets']

queries_for_ids = deque([])
queries_for_match_data = deque([])

matches_db = db.matches
champions_db = db.champions

match_count = 0.0
resting = False


regions = [
	{'endpoint': 'https://br.api.pvp.net', 'name': 'br'},
	{'endpoint': 'https://na.api.pvp.net', 'name': 'na'},
	{'endpoint': 'https://las.api.pvp.net', 'name': 'las'},
	{'endpoint': 'https://lan.api.pvp.net', 'name': 'lan'},
	{'endpoint': 'https://oce.api.pvp.net', 'name': 'oce'},
	{'endpoint': 'https://eune.api.pvp.net', 'name': 'eune'}
]

def existsInArray(arr, element, key):
	for elem in arr:
		if elem[key] == element[key]:
			return True
	return False

while True:
	print 'Starting loop'
	if len(queries_for_match_data) >= 20:
		resting = True
	else:
		resting = False

	if not resting:
		for region in regions:
			query = {
				'query': '%s/api/lol/%s/v4.1/game/ids?beginDate=%s&api_key=%s' % (region['endpoint'], region['name'], CURRENTTIME, APITOKEN),
				'region': region['name'],
				'endpoint': region['endpoint']
			}

			if not existsInArray(queries_for_ids, query, 'query'):
				queries_for_ids.append(query)

		CURRENTTIME += TIME_JUMP
		if len(queries_for_ids) == 0:
			continue


		currentQuery = queries_for_ids.popleft()

		print "Querying: %s" % currentQuery['query'] 
		try:
			r = requests.get(currentQuery['query'])
		except:
			CURRENTTIME -= TIME_JUMP
			time.sleep(SLEEP_TIME)
			continue
	
		if r.status_code == 429:
			queries_for_ids.appendleft(currentQuery)
			print 'Rate Exceeded, sleeping...'
			time.sleep(SLEEP_TIME)
			continue

		if r.status_code == 400:
			queries_for_ids.appendleft(currentQuery)
			CURRENTTIME -= TIME_JUMP
			print 'Time frame met, sleeping...'
			time.sleep(SLEEP_TIME)
			continue

	if resting or r.status_code == 200 or r.status == 400:
		if not resting and r.status_code == 200:
			for id in json.loads(r.content):
				query = {
					'query': '%s/api/lol/%s/v2.2/match/%s?includeTimeline=false&api_key=%s' % (currentQuery['endpoint'], currentQuery['region'], id, APITOKEN),
					'region': currentQuery['region']
				}
				if not existsInArray(queries_for_match_data, query, 'query'):
					queries_for_match_data.append(query)

		if not resting and r.status_code == 400:
			CURRENTTIME -= TIME_JUMP
			queries_for_ids.appendleft(currentQuery)

		if len(queries_for_match_data) == 0:
			continue

		currentQuery = queries_for_match_data.popleft()

		try:
			print "Querying: %s" % currentQuery['query']
			r = requests.get(currentQuery['query'])
		except:
			queries_for_match_data.appendleft(currentQuery)
			time.sleep(SLEEP_TIME)
			continue

		if r.status_code == 429:
			queries_for_match_data.appendleft(currentQuery)
			print 'Rate exceeded, sleeping...'
			time.sleep(SLEEP_TIME)
			continue

		data = json.loads(r.content)
		match = {
			'id': data['matchId'],
			'region': data['region'].lower(),
			'championsWin': [],
			'championsLose': []
		}

		for team in data['teams']:
			if team['winner'] == True:
				match['winnerTeamId'] = team['teamId']

		for participant in data['participants']:
			if participant['teamId'] == match['winnerTeamId']:
				match['championsWin'].append(participant['championId'])
			else:
				match['championsLose'].append(participant['championId'])

		print 'match data to save...'
		print match
		matches_db.insert_one(match)
		match_count += 1.0

		used_champs = []

		for champ in match['championsWin']:
			try:
				used_champs.index(champ)
			except ValueError:
				champ_data = champions_db.find_one({'id': champ})
				champ_data['totalAppearances'] += 1
				champ_data['appearanceRatio'] = (champ_data['totalAppearances']*1.0) / match_count
				champions_db.update_one({'id': champ}, {'$set': {'appearanceRatio': champ_data['appearanceRatio'], 'totalAppearances': champ_data['totalAppearances']}})
				used_champs.append(champ)

		for champ in match['championsLose']:
			try:
				used_champs.index(champ)
			except ValueError:
				champ_data = champions_db.find_one({'id': champ})
				champ_data['totalAppearances'] += 1
				champ_data['appearanceRatio'] = (champ_data['totalAppearances']*1.0) / match_count
				champions_db.update_one({'id': champ}, {'$set': {'appearanceRatio': champ_data['appearanceRatio'], 'totalAppearances': champ_data['totalAppearances']}})
				used_champs.append(champ)

	else:
		print "Unknown status: %i" % t.status_code
		
